apache/superset
# [Refactor] Extend color scheme management to sequential schemes

### Motivation

Centralize color management for ease of maintenance and customization. Also good for future migration to an independent `npm` package in `superset-ui-core`.
- Color scheme metadata are scattered. The human-readable labels are defined in `controls.jsx`. The colors are in separate file. 
- Categorical color management has been improved in #5815 , but sequential color scheme remains a single giant file with no easy way to add/remove schemes. 

### Changes

- Create data structure `ColorScheme` and children `CategoricalScheme` and `SequentialScheme` to hold `colors` array, and its metadata (`name`, `label`, `description`, `isDiverging`) in one place. Convert the arrays of colors in `colorSchemes` directory and `sequential.js` to use these new data structures.
- Leverage `Registry` class that was introduced with the chart plugin system. `ColorSchemeManager` is renamed to `ColorSchemeRegistry` and extends from `Registry`. 
- Add new function `getMap`, `getMapAsPromise`, `values()`, `valuesAsPromise()` and `clear` to `Registry` with new unit tests.
- Create two children singleton classes `CategoricalSchemeManager` and `SequentialSchemeManager` to manage categorical and sequential scheme, respectively.
- Update usage and corresponding unit tests throughout the app.

- Extract color setup from `common.js` into `setupColors.js`
- Move color-related modules under `src/modules/colors` directory and unit tests under `spec/modules/colors`.

### Test plan

- New unit tests are added. Existing unit tests are updated. 
- Verify color picker controls locally that it works like before. 
- Verify visualizations locally that colors are the same as before. 

@williaster @conglei @graceguo-supercat @michellethomas
